agent/src/agentclient.rs

```diff
@@ -20,90 +25,139 @@ use models::{
     },
 };
 
+// The reflector uses exponential backoff.
+// These values configure how long to delay between tries.
+const RETRY_BASE_DELAY: Duration = Duration::from_millis(1000);
+const RETRY_MAX_DELAY: Duration = Duration::from_secs(10);
+const NUM_RETRIES: usize = 5;
+
 const AGENT_SLEEP_DURATION: Duration = Duration::from_secs(5);
 
 #[derive(Clone)]
 pub struct BrupopAgent<T: APIServerClient> {
     k8s_client: kube::client::Client,
-    node_selector: Option<BottlerocketNodeSelector>,
     apiserver_client: T,
+    brn_reader: Store<BottlerocketNode>,
+    node_reader: Store<Node>,
+    associated_node_name: String,
+    associated_bottlerocketnode_name: String,
 }
 
 impl<T: APIServerClient> BrupopAgent<T> {
-    pub fn new(k8s_client: kube::client::Client, apiserver_client: T) -> Self {
+    pub fn new(
+        k8s_client: kube::client::Client,
+        apiserver_client: T,
+        brn_reader: Store<BottlerocketNode>,
+        node_reader: Store<Node>,
+        associated_node_name: String,
+        associated_bottlerocketnode_name: String,
+    ) -> Self {
         BrupopAgent {
             k8s_client,
             apiserver_client,
-            node_selector: None,
+            brn_reader,
+            node_reader,
+            associated_node_name,
+            associated_bottlerocketnode_name,
         }
     }
 
     pub async fn check_node_custom_resource_exists(&mut self) -> Result<bool> {
```

Comment:
These two name`check_custom_resource_status_exists` and `check_node_custom_resource_exists` are really confusing for me.
Seems like they use different `associated_bottlerocketnode_name` to fetch the customer resources from Kubernetes. `check_node_custom_resource_exists ` use the one binded with agent, `check_custom_resource_status_exists` use something related to "brn-{}" with value get from node selector, and the node_selector fetches stuff from `node_reader`.

Could you please explain more so I could better understand?

1. What is the difference between `brn_reader` and `node_reader`?
From my understand `brn_reader`is related to the customer resource you defined for monitor the status for each kubernetes node. Like if brupop is going to monitor 3 kubernetes hosts in the cluster, it will have 3 `BottlerocketNode` customer resources so the agent and controller can work together to update the status and spec to upgrade the host.
But I don't understand the `node_reader` here, what does that do? 

2. Could you explain the differences between `check_custom_resource_status_exists` and `check_node_custom_resource_exists`? And maybe choose a better name or add more comments on it?